crates/bevy_asset/src/io/metadata.rs

```diff
@@ -0,0 +1,64 @@
+use std::convert::{TryFrom, TryInto};
+
+/// A enum representing a type of file.
+#[non_exhaustive]
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub enum FileType {
+    Directory,
+    File,
+}
+
+impl FileType {
+    pub const fn is_dir(&self) -> bool {
+        matches!(self, Self::Directory)
+    }
+
+    pub const fn is_file(&self) -> bool {
+        matches!(self, Self::File)
+    }
+}
+
+impl TryFrom<std::fs::FileType> for FileType {
+    type Error = std::io::Error;
+
+    fn try_from(file_type: std::fs::FileType) -> Result<Self, Self::Error> {
+        if file_type.is_dir() {
+            Ok(Self::Directory)
+        } else if file_type.is_file() {
+            Ok(Self::File)
+        } else {
+            Err(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "unknown file type",
+            ))
+        }
+    }
+}
+
+/// Metadata information about a file.
+///
+/// This structure is returned from the [`AssetIo::get_metadata`] method.
+#[derive(Debug, Clone)]
+pub struct Metadata {
+    pub file_type: FileType,
+}
+
+impl Metadata {
+    pub const fn is_dir(&self) -> bool {
```

Comment:
```suggestion
    #[inline]
    pub const fn is_dir(&self) -> bool {
```

Reply:
It's actually probably worth going through and marking most of these functions `#[inline]`